devito/function.py

```diff
@@ -1074,3 +1122,107 @@ def __indices_setup__(cls, **kwargs):
     @classmethod
     def __shape_setup__(cls, **kwargs):
         return kwargs.get('shape', (kwargs.get('nt'), kwargs.get('npoint'),))
+
+
+class PrecomputedSparseFunction(AbstractSparseFunction):
+    is_PrecomputedSparseFunction = True
+
+    def __init__(self, *args, **kwargs):
+        if not self._cached():
+            super(PrecomputedSparseFunction, self).__init__(*args, **kwargs)
+
+            # Grid points per sparse point (2 in the case of bilinear and trilinear)
+            r = kwargs.get('r')
+            if not isinstance(r, int) and r > 0:
+                raise ValueError('Interpolation requires parameter `r` (>0)')
+            self.r = r
+
+            gridpoints = Function(name="%s_gridpoints" % self.name,
+                                  dimensions=(self.indices[-1], Dimension(name='d')),
+                                  shape=(self.npoint, self.grid.dim), space_order=0,
+                                  dtype=np.int32)
+
+            gridpoints_data = kwargs.get('gridpoints', None)
+            assert(gridpoints_data is not None)
+            gridpoints.data[:] = gridpoints_data[:]
+            self.gridpoints = gridpoints
+
+            coefficients = Function(name="%s_coefficients" % self.name,
+                                    dimensions=(self.indices[-1], Dimension(name='d'),
+                                                Dimension(name='i')),
+                                    shape=(self.npoint, self.grid.dim, self.r),
+                                    space_order=0)
+            coefficients_data = kwargs.get('coefficients', None)
```

Comment:
So for gridpoints/coefficients, wouldn't it be more consistent to keep `coordinate` and have an `interpolator` as input that computes gridpoints and coefficients from the coordinates.
```
sf = PrecomputedSparseFunction(name='s', grid=grid, r=r, npoint=len(points),
                                                   coordinates=coordinate, inteprolator=precompute_linear_interpolation)
```

Reply:
You also need grid spacing and origin to map from coordinates to gridpoints. Can't stay consistent if we have to add those values too.

Reply:
Why make this one the only Function that does not take a grid in?

Reply:
It [does](https://github.com/opesci/devito/pull/515/files#diff-f3329261db6f5e1efaecf750bcfc52e9R95)?

Reply:
If it follow every other function, you have a Grid, so you have spacing and origin

Reply:
The grid only has *symbols* for spacing and origin which are replaced by values only at apply time. To do any precomputation, I need those *values* before apply time. Not possible with the grid object.

Reply:
> I need those values before apply time.

You shouldn't , the `subs` is supposed to take care of it. Hard codingh values that may have nothing to do with the one that will be used at runtime is super risky.

Reply:
And Grid always have spacing/origin values

Reply:
Combining 
> You also need grid spacing and origin to map from coordinates to gridpoints.
with the fact that gridpoints need to be **pre**computed means I need spacing and origin before apply time. I don't know how else to explain this.
Hence, it's best to keep this mess outside of our classes and in user space - which is where this PR keeps it.

Reply:
I find this really hacky and error prone.

Reply:
This was raised in a meeting and implemented after agreement with you, @ggorman and @FabioLuporini .

Reply:
Forcing the user to do the precomputation is better than silently using the origin and spacing values used to intialise the grid - since these might be changed at apply time, following your own logic set out in:
>Hard codingh values that may have nothing to do with the one that will be used at runtime is super risky.

Reply:
Makes sense. Can we add some warnings at setup/runtime saying the setup is only valid for a specific grid?

This all PR is gonna be really handy for sparse-fd btw, so not trying to stop it by anyway, just making sure we don't making too different than the rest of SparseFunctions

Reply:
Sure.